plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeSplitManager.java

```diff
@@ -177,7 +179,9 @@ private Stream<DeltaLakeSplit> getSplits(
 
         MetadataEntry metadataEntry = tableHandle.getMetadataEntry();
         boolean isOptimize = tableHandle.isOptimize();
-        Set<Map<String, Optional<String>>> partitionsWithAtMostOneFile = isOptimize ? findPartitionsWithAtMostOneFile(validDataFiles) : ImmutableSet.of();
+        // Materialize the valid data files iterator to a list only in case of dealing with OPTIMIZE to avoid rewriting partitions containing only one file
+        List<AddFileEntry> validDataFilesList = isOptimize ? ImmutableList.copyOf(validDataFilesIterator) : ImmutableList.of();
+        Set<Map<String, Optional<String>>> partitionsWithAtMostOneFile = isOptimize ? findPartitionsWithAtMostOneFile(validDataFilesList) : ImmutableSet.of();
```

Comment:
You could change argument of `findPartitionsWithAtMostOneFile` to Stream and get rid of List